### DISCUSSION TITLE
Converting coordinates from TEME frame to ICRF frame

### DISCUSSION THREAD

**Discussion by W-Hunter-Daboll** [skyfielders/python-skyfield#723] [Q&A] [answered]
I'm looking to convert data from the TEME (true equator, mean equinox) frame to the ICRF (International Celestial Reference System) frame.  This data has already been converted from TLE state data into cartesian xyz position and uvw velocity in the TEME frame using the python [sgp4 ](https://pypi.org/project/sgp4/) package. 

I understand that converting TEME to ICRF can be done with the skyfield package, and that historically this might be done using the TEME_to_ITRF function under [skyfield/sgp4lib.py](https://github.com/skyfielders/python-skyfield/blob/master/skyfield/sgp4lib.py). However, that function has been deprecated in favor of the TEME and ITRS frame objects (see referenced code in image below). 

![TEME_to_ICRF_Function](https://user-images.githubusercontent.com/69279360/158431787-49936e51-fe9f-463b-910c-5a73b75be6ff.JPG)

Unfortunately, I'm not easily finding these frame objects. If someone could point me to their location and the relevant documentation to implement this coordinate frame rotation, I'd be grateful. 

Thanks!

P.S. Other relevant research I've done:
I know from the [Skyfield documentation on Earth Satellites](https://rhodesmill.org/skyfield/earth-satellites.html) that it is possible to read a TLE into skyfield directly, and that this will be in the Geocentric Celestial Reference System (GCRS) frame. Unfortunately, this is not useful to me as I need the coordinates in ICRF. 

I also know that [Skyfield internally stores all position data in the ICRF frame](https://rhodesmill.org/skyfield/positions.html) - I find myself not knowing how to input position in TEME and get ICRF back out. 
![ICRF_Internal](https://user-images.githubusercontent.com/69279360/158433703-1b1aea77-6676-49ac-b8ff-6e2ad8e410b5.JPG)

**Comment by brandon-rhodes** [comment, marked as answer]
Good question. The conversion would follow the conventions laid out in the Coordinates document:

https://rhodesmill.org/skyfield/coordinates.html

The TEME frame can probably be imported like this:

```
from skyfield.sgp4lib import TEME
```

Probably I should expand the Coordinates document to at least mention TEME and show where it can be imported from.

**Reply by W-Hunter-Daboll** [reply]
That's a great help, thank you! 

I can see roughly how I would go about starting with a position in ICRS and getting position in TEME out, although I don't quite have the syntax right on this first try: 
![image](https://user-images.githubusercontent.com/69279360/158462136-aad2ca64-9803-4cd9-aa79-7de3bafdb255.png)

But how would I instead input TEME position and get out ICRS? I don't see an easy way to declare a TEME position using the build_position or Apparent.from_radec functions, as are exemplified in the Coordinates document. 
![image](https://user-images.githubusercontent.com/69279360/158462456-3c53a648-36b6-4f25-aa5c-6ea32e6641a9.png)

Apologies if my question is uniformed, and thank you for your swift reply!

**Reply by W-Hunter-Daboll** [reply]
*Apologies if my question is uninformed ...

**Reply by brandon-rhodes** [reply]
Take a look at the [from_time_and_frame_vectors()](https://rhodesmill.org/skyfield/api-position.html#skyfield.positionlib.ICRF.from_time_and_frame_vectors) method mentioned briefly at the bottom of that section, I think you can give it TEME coordinates and have an ICRS position come out.

**Reply by W-Hunter-Daboll** [reply]
Thank you!
